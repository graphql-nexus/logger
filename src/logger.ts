import * as Lo from 'lodash'
import * as OS from 'os'
import { Level, LevelNum, LEVELS } from './level'
import * as Prettifier from './prettifier'
import * as RootLogger from './root-logger'

// TODO JSON instead of unknown type
type Context = Record<string, unknown>

export type LogRecord = {
  level: LevelNum
  path: string[]
  event: string
  context: Context
  time: number
  pid: number
  hostname: string
}

type Log = (event: string, context?: Context) => void

export type Logger = {
  fatal: Log
  error: Log
  warn: Log
  info: Log
  debug: Log
  trace: Log
  addToContext: (context: Context) => Logger // fluent
  child: (name: string) => Logger // fluent
}

/**
 * Create a logger.
 */
export function create(
  rootState: RootLogger.State,
  path: string[],
  parentContext: Context
): { logger: Logger; link: Link } {
  const state: State = {
    // Copy as addToContext will mutate it
    pinnedAndParentContext: Lo.cloneDeep(parentContext),
    children: [],
  }

  function updateContextAndPropagate(newContext: Context) {
    state.pinnedAndParentContext = newContext
    state.children.forEach((child) => {
      child.onNewParentContext(state.pinnedAndParentContext)
    })
  }

  function send(level: Level, event: string, localContext: undefined | Context) {
    const thisLevel = LEVELS[level].number
    const levelSetting = LEVELS[rootState.settings.level].number
    if (thisLevel >= levelSetting) {
      // Avoid mutating the passed local context
      const context = localContext
        ? Lo.merge({}, state.pinnedAndParentContext, localContext)
        : state.pinnedAndParentContext
      const logRec: LogRecord = {
        path,
        context,
        event,
        level: thisLevel,
        time: Date.now(),
        hostname: OS.hostname(),
        pid: process.pid,
      }
      const logMsg = rootState.settings.pretty.enabled
        ? Prettifier.render(rootState.settings.pretty, logRec)
        : JSON.stringify(logRec)
      rootState.settings.output.write(logMsg + OS.EOL)
    }
  }

  const link: Link = {
    onNewParentContext(newParentContext: Context) {
      updateContextAndPropagate(
        Lo.merge(
          // Copy so that we don't mutate parent while maintaining local overrides...
          {},
          newParentContext,
          // ...this
          state.pinnedAndParentContext
        )
      )
    },
  }

  const logger: Logger = {
    fatal(event, context) {
      send('fatal', event, context)
    },
    error(event, context) {
      send('error', event, context)
    },
    warn(event, context) {
      send('warn', event, context)
    },
    info(event, context) {
      send('info', event, context)
    },
    debug(event, context) {
      send('debug', event, context)
    },
    trace(event, context) {
      send('trace', event, context)
    },
    addToContext(context: Context) {
      // Can safely mutate here, save some electricity...
      updateContextAndPropagate(Lo.merge(state.pinnedAndParentContext, context))
      return logger
    },
    child: (name: string): Logger => {
      const { logger: child, link } = create(rootState, path.concat([name]), state.pinnedAndParentContext)
      state.children.push(link)
      return child
    },
  }

  return {
    logger,
    link,
  }
}

type Link = {
  onNewParentContext: (newContext: Context) => void
}

type State = {
  pinnedAndParentContext: Context
  children: Link[]
}
