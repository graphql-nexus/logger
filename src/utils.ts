import { format } from 'util'
/**
 * Guarantee the length of a given string, padding before or after with the
 * given character. If the given string is longer than  the span target, then it
 * will be cropped.
 */
export function span(
  padSide: 'padBefore' | 'padAfter',
  padChar: string,
  target: number,
  content: string
): string {
  if (content.length > target) {
    return content.slice(0, target)
  }
  let toPadSize = target - content.length
  while (toPadSize > 0) {
    if (padSide === 'padAfter') {
      content = content + padChar
    } else if (padSide === 'padBefore') {
      content = padChar + content
    }
    toPadSize--
  }
  return content
}

/**
 * Guarantee the length of a given string, padding with space as needed. Content
 * is aligned left and if exceeds span target length to begin with gets cropped.
 */
export const clampSpace = span.bind(null, 'padAfter', ' ')

/**
 * Create a string of space of the given length.
 */
export function spanSpace(num: number): string {
  return spanChar(num, ' ')
}

/**
 * Create a string of the given length and character
 */
export function spanChar(num: number, char: string): string {
  return range(num).map(constant(char)).join('')
}

/**
 * Guarantee the length of a given string, padding with space as needed. Content
 * is aligned right and if exceeds span target length to begin with gets cropped.
 */
export const spanSpaceRight = span.bind(null, 'padBefore', ' ')

/**
 * Use this to make assertion at end of if-else chain that all members of a
 * union have been accounted for.
 */
export function casesHandled(x: never): never {
  throw new Error(`A case of value was not handled: ${x}`)
}

/**
 * Create a function that will only ever return the given value when called.
 */
export function constant<T>(x: T): () => T {
  return function () {
    return x
  }
}

/**
 * Create a range of integers.
 */
export function range(times: number): number[] {
  const list: number[] = []
  while (list.length < times) {
    list.push(list.length + 1)
  }
  return list
}

/**
 * Strip keys from object whose value is undefined.
 */
export function omitUndefinedKeys<T extends Record<string, unknown>>(data: T): T {
  return Object.entries(data ?? {})
    .filter(([k, v]) => v !== undefined)
    .reduce((acc, [k, v]) => Object.assign(acc, { [k]: v }), {} as T)
}

/**
 * Get the last item of an array.
 */
export function last<T>(xs: T[]): T {
  return xs[xs.length - 1]
}

export function isEmpty(x?: object): boolean {
  if (x === undefined) return true
  return Object.values(x).filter((val) => val !== undefined).length === 0
}

/**
 * Run a given parser over an environment variable. If parsing fails, throw a
 * contextual error message.
 */
export function parseFromEnvironment<T>(
  key: string,
  parser: {
    info: { valid: string; typeName: string }
    run: (raw: string) => null | T
  }
): T {
  const envVarValue = process.env[key]! // assumes env presence handled before
  const result = parser.run(envVarValue)

  if (result === null) {
    throw new Error(
      `Could not parse environment variable ${key} into ${
        parser.info.typeName
      }. The environment variable was: ${format(envVarValue)}. A valid environment variable must be like: ${
        parser.info.valid
      }`
    )
  }

  return result
}
