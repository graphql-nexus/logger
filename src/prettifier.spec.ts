import * as OS from 'os'
import { chalk } from './chalk'
import { LogRecord } from './logger'
import * as Prettifier from './prettifier'
import { spanChar } from './utils'

function makeRec(data?: Omit<Partial<LogRecord>, 'event'>): LogRecord {
  return {
    hostname: 'host',
    pid: 0,
    time: 0,
    level: 10,
    path: ['root'],
    context: {},
    ...data,
    event: 'foo', // do not allow changing b/c headers width below depends on this value
  }
}

function render(): string {
  return Prettifier.render(options, rec)
}

let options: Prettifier.Options
let rec: LogRecord
let terminalWidth: number
let terminalContextWidth: number

beforeEach(() => {
  let logHeadersWidth = ('● root foo' + Prettifier.separators.context.singleLine.symbol).length
  chalk.level = 0 // disable color
  options = {
    color: false, // only disables color of util.inspect
    levelLabel: false,
    timeDiff: false,
  }
  rec = makeRec()
  terminalWidth = 100
  terminalContextWidth = terminalWidth - logHeadersWidth
  process.stdout.columns = terminalWidth
})

describe('singleline', () => {
  it('used if context does fit singleline', () => {
    rec.context = stringValueEntryWithin('key', terminalContextWidth)
    const l = render()
    expect(l).toMatchSnapshot()
    expect(trimTrailingNewline(l).length).toBeLessThanOrEqual(terminalWidth)
  })
  it('used if context does fit singleline (multiple key-values)', () => {
    rec.context = {
      ...stringValueEntryWithin('ke1', terminalContextWidth / 2),
      ...stringValueEntryWithin(
        'ke2',
        terminalContextWidth / 2 - Prettifier.separators.contextEntry.singleLine.length
      ),
    }
    const l = render()
    expect(l).toMatchSnapshot()
    expect(trimTrailingNewline(l).length).toBeLessThanOrEqual(terminalWidth)
  })
  // it('objects are formatted by util.inspect compact: yes', () => {
  //   if (!process.version.match(/^v12/)) return
  //   log.info('foo', { ke1: { a: { b: { c: true } } } })
  //   expect(output.memory.jsonOrRaw).toMatchSnapshot()
  // })
})

describe('multiline', () => {
  it('used if context does not fit singleline', () => {
    rec.context = stringValueEntryWithin('key', terminalContextWidth + 1 /* force multi */)
    expect(render()).toMatchSnapshot()
  })

  it('used if context does fit singleline (multiple key-values)', () => {
    rec.context = {
      ...stringValueEntryWithin('ke1', terminalContextWidth / 2),
      ...stringValueEntryWithin(
        'ke2',
        terminalContextWidth / 2 - Prettifier.separators.contextEntry.singleLine.length + 1 /* force multi */
      ),
    }
    expect(render()).toMatchSnapshot()
  })
  // it('objects are formatted by util.inspect compact: yes', () => {
  //   if (!process.version.match(/^v12/)) return
  //   log.info('foo', {
  //     ke1: {
  //       a: {
  //         b: {
  //           c: true,
  //           d: 'looooooooooooooooooooooooooooooooooooooooooooooooong',
  //         },
  //       },
  //     },
  //   })
  //   expect(output.memory.jsonOrRaw).toMatchSnapshot()
  // })
})

// helpers for building content for tests against log formatting

function stringValueWithin(size: number): string {
  const actualSize = size - 2 // -2 for quote rendering "'...'"
  const value = spanChar(actualSize, 'x')
  return value
}

function stringValueEntryWithin(keyName: string, size: number): Record<any, any> {
  const KeyWidth = keyName.length + Prettifier.separators.contextKeyVal.singleLine.symbol.length
  return {
    [keyName]: stringValueWithin(size - KeyWidth),
  }
}

/**
 * Remove traiing newline. Strict alternative to .trim().
 */
function trimTrailingNewline(s: string): string {
  return s.replace(new RegExp(OS.EOL + '$'), '')
}
