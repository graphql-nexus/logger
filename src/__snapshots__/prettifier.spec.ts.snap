// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiline used if context does fit singleline (multiple key-values) 1`] = `
"— root foo
  | ke1  'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
  | ke2  'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'"
`;

exports[`multiline used if context does not fit singleline 1`] = `
"— root foo
  | key  'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'"
`;

exports[`singleline used if context does fit singleline (multiple key-values) 1`] = `"— root foo  --  ke1: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'  ke2: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'"`;

exports[`singleline used if context does fit singleline 1`] = `"— root foo  --  key: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'"`;
